### PR TITLE
fix: always put to a temp location for download and extract

### PR DESCRIPTION
**Pull request by unknwon**
If the process gets killed halfway through download or extraction, we will get incomplete archives or binaries.

This PR changes the process first to a temporary location, and only moves to the target location once the download and extraction processes are finished.

## Test plan

```
$ rm -rf ~/.embedded-postgres-go
$ rm -rf ~/Library/Application\ Support/sourcegraph-sp/postgresql/bin
$ ./sourcegraph
```

---

Part of https://github.com/sourcegraph/sourcegraph/issues/48893

### CODE REVIEW

File: embedded_postgres.go
Comment (keegancsmith) [inline embedded_postgres.go:113]: additionally add a defer to remove the temp location so we clean up if decompressTarXz or os.Rename fail.

File: remote_fetch.go
Comment (keegancsmith) [inline remote_fetch.go:98]: don't you also need that extra deletion logic like you have above?

**Review by keegancsmith — APPROVED** [review]
approving to unblock. But to do this durably I'd copy over this function https://sourcegraph.com/github.com/sourcegraph/sourcegraph@6fd6a81f60711719b366a234be03850a6378dde0/-/blob/internal/fileutil/fileutil.go?L59

**Review by emidoots — APPROVED** [review]

File: remote_fetch.go
Comment (unknwon) [reply remote_fetch.go:98]: I kinda rely on the truncate behvaior of `os.WriteFile` when the file already exists on path, but doesn't hurt to do this explicitly.

**Comment by unknwon** [comment]
Merging as-is, happy to address any post-merge comments!

**Review by keegancsmith — COMMENTED** [review]
LGTM

**Pull request merged**
